core/src/main/java/org/apache/iceberg/rest/RESTTableOperations.java

```diff
@@ -169,16 +181,21 @@ public FileIO io() {
     return io;
   }
 
-  private TableMetadata updateCurrentMetadata(LoadTableResponse response) {
-    // LoadTableResponse is used to deserialize the response, but config is not allowed by the REST
-    // spec so it can be
-    // safely ignored. there is no requirement to update config on refresh or commit.
-    if (current == null
-        || !Objects.equals(current.metadataFileLocation(), response.metadataLocation())) {
-      this.current = response.tableMetadata();
+  private void updateCurrentMetadata(LoadTableResponse response) {
+    this.current = response.tableMetadata();
+  }
+
+  private void updateIO(LoadTableResponse response) {
+    Map<String, String> mergedConfig = RESTUtil.merge(this.io.properties(), response.config());
+    String ioImpl =mergedConfig.getOrDefault(
+            CatalogProperties.FILE_IO_IMPL, CatalogProperties.DEFAULT_REST_FILE_IO_IMPL);
+    Object hadoopConf = null;
+    if (this.io instanceof HadoopConfigurable) {
+      hadoopConf = ((HadoopConfigurable) this.io).getConf();
     }
 
-    return current;
+    closeableGroup.addCloseable(this.io);
+    this.io = CatalogUtil.loadFileIO(ioImpl, mergedConfig, hadoopConf);
```

Comment:
What about tracking the FileIO? How does this signal to the parent that the FileIO should be closed when the parent is closed? What closes this class?

I also think this needs to detect changes and not just replace the FileIO instance every refresh. That's wasteful and creates a lot of objects that need to be tracked and closed.

Reply:
Agreed, I think we need to pick one of the approaches @RussellSpitzer described [above](https://github.com/apache/iceberg/pull/11194#discussion_r1783423609) in order to properly manage the lifecycle of the FileIO instances and also to support the bifunctor in the `Catalog`.